### ISSUE TITLE
ADOT layer causes TypeError when using --warm-containers feature with start-api

### ISSUE THREAD

**Issue by hooverdc** [aws/aws-sam-cli#3683]
For reference: https://aws-otel.github.io/docs/getting-started/lambda/lambda-python

### Description:

When using the AWS managed Lambda layer for ADOT and the sam start-api with `--warm-containers`, the Lambda container will fail to initialize.

### Steps to reproduce:

Require layer `arn:aws:lambda:us-west-2:901920570463:layer:aws-otel-python38-amd64-ver-1-9-1:1` in your SAM function definition. Call `sam local start-api --warm-containers EAGER|LAZY`. Unable to replicate without the `--warm-containers` flag.

### Observed result:

```bash
2022-02-22 17:05:10,378 | Telemetry endpoint configured to be https://aws-serverless-tools-telemetry.us-west-2.amazonaws.com/metrics
2022-02-22 17:05:10,378 | Using config file: samconfig.toml, config environment: default
2022-02-22 17:05:10,378 | Expand command line arguments to:
2022-02-22 17:05:10,378 | --template_file=/Users/hooverd/Projects/account-api-rest/.aws-sam/build/template.yaml --env_vars=/dev/fd/63 --warm_containers=LAZY --host=127.0.0.1 --port=3000 --static_dir=public --layer_cache_basedir=/Users/hooverd/.aws-sam/layers-pkg --container_host=localhost --container_host_interface=127.0.0.1 
2022-02-22 17:05:10,489 | local start-api command is called
2022-02-22 17:05:10,513 | Collected default values for parameters: {}
2022-02-22 17:05:10,541 | There is no customer defined id or cdk path defined for resource AccountApi, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,541 | There is no customer defined id or cdk path defined for resource AuthorizerFunctionLambdaPermission, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,541 | Sam customer defined id is more priority than other IDs. Customer defined id for resource AuthorizerFunction is AuthorizerFunction
2022-02-22 17:05:10,541 | There is no customer defined id or cdk path defined for resource AccountApiSecurityGroup, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,541 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SharedLayer is SharedLayer
2022-02-22 17:05:10,541 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetPersonInfoFunction is GetPersonInfoFunction
2022-02-22 17:05:10,541 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetAccountInfoFunction is GetAccountInfoFunction
2022-02-22 17:05:10,544 | 7 stacks found in the template
2022-02-22 17:05:10,544 | Collected default values for parameters: {}
2022-02-22 17:05:10,560 | There is no customer defined id or cdk path defined for resource AccountApi, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,561 | There is no customer defined id or cdk path defined for resource AuthorizerFunctionLambdaPermission, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,561 | Sam customer defined id is more priority than other IDs. Customer defined id for resource AuthorizerFunction is AuthorizerFunction
2022-02-22 17:05:10,561 | There is no customer defined id or cdk path defined for resource AccountApiSecurityGroup, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,561 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SharedLayer is SharedLayer
2022-02-22 17:05:10,561 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetPersonInfoFunction is GetPersonInfoFunction
2022-02-22 17:05:10,561 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetAccountInfoFunction is GetAccountInfoFunction
2022-02-22 17:05:10,563 | 7 resources found in the stack 
2022-02-22 17:05:10,563 | Collected default values for parameters: {}
2022-02-22 17:05:10,579 | There is no customer defined id or cdk path defined for resource AccountApi, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,579 | There is no customer defined id or cdk path defined for resource AuthorizerFunctionLambdaPermission, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,579 | Sam customer defined id is more priority than other IDs. Customer defined id for resource AuthorizerFunction is AuthorizerFunction
2022-02-22 17:05:10,579 | There is no customer defined id or cdk path defined for resource AccountApiSecurityGroup, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,579 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SharedLayer is SharedLayer
2022-02-22 17:05:10,579 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetPersonInfoFunction is GetPersonInfoFunction
2022-02-22 17:05:10,579 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetAccountInfoFunction is GetAccountInfoFunction
2022-02-22 17:05:10,581 | Collected default values for parameters: {}
2022-02-22 17:05:10,598 | There is no customer defined id or cdk path defined for resource AccountApi, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,598 | There is no customer defined id or cdk path defined for resource AuthorizerFunctionLambdaPermission, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,598 | Sam customer defined id is more priority than other IDs. Customer defined id for resource AuthorizerFunction is AuthorizerFunction
2022-02-22 17:05:10,598 | There is no customer defined id or cdk path defined for resource AccountApiSecurityGroup, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,598 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SharedLayer is SharedLayer
2022-02-22 17:05:10,598 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetPersonInfoFunction is GetPersonInfoFunction
2022-02-22 17:05:10,598 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetAccountInfoFunction is GetAccountInfoFunction
2022-02-22 17:05:10,600 | --base-dir is not presented, adjusting uri SharedLayer relative to /Users/hooverd/Projects/account-api-rest/.aws-sam/build/template.yaml
2022-02-22 17:05:10,600 | Found Serverless function with name='AuthorizerFunction' and CodeUri='AuthorizerFunction'
2022-02-22 17:05:10,600 | --base-dir is not presented, adjusting uri AuthorizerFunction relative to /Users/hooverd/Projects/account-api-rest/.aws-sam/build/template.yaml
2022-02-22 17:05:10,600 | Collected default values for parameters: {}
2022-02-22 17:05:10,617 | There is no customer defined id or cdk path defined for resource AccountApi, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,617 | There is no customer defined id or cdk path defined for resource AuthorizerFunctionLambdaPermission, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,617 | Sam customer defined id is more priority than other IDs. Customer defined id for resource AuthorizerFunction is AuthorizerFunction
2022-02-22 17:05:10,617 | There is no customer defined id or cdk path defined for resource AccountApiSecurityGroup, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,617 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SharedLayer is SharedLayer
2022-02-22 17:05:10,617 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetPersonInfoFunction is GetPersonInfoFunction
2022-02-22 17:05:10,617 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetAccountInfoFunction is GetAccountInfoFunction
2022-02-22 17:05:10,619 | --base-dir is not presented, adjusting uri SharedLayer relative to /Users/hooverd/Projects/account-api-rest/.aws-sam/build/template.yaml
2022-02-22 17:05:10,619 | Found Serverless function with name='GetPersonInfoFunction' and CodeUri='GetPersonInfoFunction'
2022-02-22 17:05:10,619 | --base-dir is not presented, adjusting uri GetPersonInfoFunction relative to /Users/hooverd/Projects/account-api-rest/.aws-sam/build/template.yaml
2022-02-22 17:05:10,619 | Collected default values for parameters: {}
2022-02-22 17:05:10,635 | There is no customer defined id or cdk path defined for resource AccountApi, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,635 | There is no customer defined id or cdk path defined for resource AuthorizerFunctionLambdaPermission, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,635 | Sam customer defined id is more priority than other IDs. Customer defined id for resource AuthorizerFunction is AuthorizerFunction
2022-02-22 17:05:10,635 | There is no customer defined id or cdk path defined for resource AccountApiSecurityGroup, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,635 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SharedLayer is SharedLayer
2022-02-22 17:05:10,635 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetPersonInfoFunction is GetPersonInfoFunction
2022-02-22 17:05:10,635 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetAccountInfoFunction is GetAccountInfoFunction
2022-02-22 17:05:10,637 | --base-dir is not presented, adjusting uri SharedLayer relative to /Users/hooverd/Projects/account-api-rest/.aws-sam/build/template.yaml
2022-02-22 17:05:10,637 | Found Serverless function with name='GetAccountInfoFunction' and CodeUri='GetAccountInfoFunction'
2022-02-22 17:05:10,637 | --base-dir is not presented, adjusting uri GetAccountInfoFunction relative to /Users/hooverd/Projects/account-api-rest/.aws-sam/build/template.yaml
2022-02-22 17:05:10,638 | watch resource /Users/hooverd/Projects/account-api-rest/.aws-sam/build/template.yaml
2022-02-22 17:05:10,638 | Create Observer for resource /Users/hooverd/Projects/account-api-rest/.aws-sam/build/template.yaml with recursive True
2022-02-22 17:05:10,640 | watch resource /Users/hooverd/Projects/account-api-rest/.aws-sam/build/template.yaml's parent /Users/hooverd/Projects/account-api-rest/.aws-sam/build
2022-02-22 17:05:10,640 | Create Observer for resource /Users/hooverd/Projects/account-api-rest/.aws-sam/build with recursive False
2022-02-22 17:05:10,649 | Collected default values for parameters: {}
2022-02-22 17:05:10,665 | There is no customer defined id or cdk path defined for resource AccountApi, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,666 | There is no customer defined id or cdk path defined for resource AuthorizerFunctionLambdaPermission, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,666 | Sam customer defined id is more priority than other IDs. Customer defined id for resource AuthorizerFunction is AuthorizerFunction
2022-02-22 17:05:10,666 | There is no customer defined id or cdk path defined for resource AccountApiSecurityGroup, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,666 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SharedLayer is SharedLayer
2022-02-22 17:05:10,666 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetPersonInfoFunction is GetPersonInfoFunction
2022-02-22 17:05:10,666 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetAccountInfoFunction is GetAccountInfoFunction
2022-02-22 17:05:10,667 | Collected default values for parameters: {}
2022-02-22 17:05:10,684 | There is no customer defined id or cdk path defined for resource AccountApi, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,684 | There is no customer defined id or cdk path defined for resource AuthorizerFunctionLambdaPermission, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,684 | Sam customer defined id is more priority than other IDs. Customer defined id for resource AuthorizerFunction is AuthorizerFunction
2022-02-22 17:05:10,684 | There is no customer defined id or cdk path defined for resource AccountApiSecurityGroup, so we will use the resource logical id as the resource id
2022-02-22 17:05:10,684 | Sam customer defined id is more priority than other IDs. Customer defined id for resource SharedLayer is SharedLayer
2022-02-22 17:05:10,684 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetPersonInfoFunction is GetPersonInfoFunction
2022-02-22 17:05:10,684 | Sam customer defined id is more priority than other IDs. Customer defined id for resource GetAccountInfoFunction is GetAccountInfoFunction
2022-02-22 17:05:10,686 | Detected Inline Swagger definition
2022-02-22 17:05:10,686 | Lambda function integration not found in Swagger document at path='/person/{uaid}' method='get'
2022-02-22 17:05:10,686 | Lambda function integration not found in Swagger document at path='/account/{id}' method='get'
2022-02-22 17:05:10,686 | Found '0' APIs in resource 'AccountApi'
2022-02-22 17:05:10,686 | Found '0' API Events in Serverless function with name 'AuthorizerFunction'
2022-02-22 17:05:10,686 | Found '1' API Events in Serverless function with name 'GetPersonInfoFunction'
2022-02-22 17:05:10,686 | Found '1' API Events in Serverless function with name 'GetAccountInfoFunction'
2022-02-22 17:05:10,686 | Removed duplicates from '2' Explicit APIs and '0' Implicit APIs to produce '2' APIs
2022-02-22 17:05:10,686 | 2 APIs found in the template
2022-02-22 17:05:10,695 | Mounting GetPersonInfoFunction at http://127.0.0.1:3000/person/{uaid} [GET]
2022-02-22 17:05:10,695 | Mounting GetAccountInfoFunction at http://127.0.0.1:3000/account/{id} [GET]
2022-02-22 17:05:10,695 | You can now browse to the above endpoints to invoke your functions. You do not need to restart/reload SAM CLI while working on your functions, changes will be reflected instantly/automatically. You only need to restart SAM CLI if you update your AWS SAM template
2022-02-22 17:05:10,695 | Localhost server is starting up. Multi-threading = True
2022-02-22 17:05:10  * Running on http://127.0.0.1:3000/ (Press CTRL+C to quit)
2022-02-22 17:05:17,607 | Constructed String representation of Event Version 2.0 to invoke Lambda. Event: {"version": "2.0", "routeKey": "GET /person/<uaid>", "rawPath": "/person/112156280046", "rawQueryString": "", "cookies": [], "headers": {"Authorization": "Bearer {{token}}", "User-Agent": "PostmanRuntime/7.29.0", "Accept": "*/*", "Cache-Control": "no-cache", "Postman-Token": "881f3879-5275-4d75-a933-d0d77fb11a1d", "Host": "127.0.0.1:3000", "Accept-Encoding": "gzip, deflate, br", "Connection": "keep-alive", "X-Forwarded-Proto": "http", "X-Forwarded-Port": "3000"}, "queryStringParameters": {}, "requestContext": {"accountId": "123456789012", "apiId": "1234567890", "http": {"method": "GET", "path": "/person/112156280046", "protocol": "HTTP/1.1", "sourceIp": "127.0.0.1", "userAgent": "Custom User Agent String"}, "requestId": "2a8373c1-f80d-4f44-8160-3e4f2f806b89", "routeKey": "GET /person/<uaid>", "stage": "$default", "time": "23/Feb/2022:00:05:10 +0000", "timeEpoch": 1645574710, "domainName": "localhost", "domainPrefix": "localhost"}, "body": "", "pathParameters": {"uaid": "xxxxxxx"}, "stageVariables": null, "isBase64Encoded": false}
2022-02-22 17:05:17,607 | Found one Lambda function with name 'GetPersonInfoFunction'
2022-02-22 17:05:17,607 | Invoking app.lambda_handler (python3.8)
2022-02-22 17:05:17,607 | Environment variables overrides data is in CloudFormation parameter file format
2022-02-22 17:05:17,607 | Loading AWS credentials from session with profile 'None'
2022-02-22 17:05:17,622 | Resolving code path. Cwd=/Users/hooverd/Projects/account-api-rest/.aws-sam/build, CodeUri=/Users/hooverd/Projects/account-api-rest/.aws-sam/build/GetPersonInfoFunction
2022-02-22 17:05:17,622 | Resolved absolute path to code is /Users/hooverd/Projects/account-api-rest/.aws-sam/build/GetPersonInfoFunction
2022-02-22 17:05:17,623 | watch resource /Users/hooverd/Projects/account-api-rest/.aws-sam/build/GetPersonInfoFunction
2022-02-22 17:05:17,623 | Create Observer for resource /Users/hooverd/Projects/account-api-rest/.aws-sam/build/GetPersonInfoFunction with recursive True
2022-02-22 17:05:17,624 | watch resource /Users/hooverd/Projects/account-api-rest/.aws-sam/build/GetPersonInfoFunction's parent /Users/hooverd/Projects/account-api-rest/.aws-sam/build
2022-02-22 17:05:17,624 | Exception on /person/112156280046 [GET]
Traceback (most recent call last):
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/samcli/local/apigw/local_apigw_service.py", line 356, in _request_handler
    self.lambda_runner.invoke(route.function_name, event, stdout=stdout_stream_writer, stderr=self.stderr)
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/samcli/commands/local/lib/local_lambda.py", line 137, in invoke
    self.local_runtime.invoke(
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/samcli/lib/telemetry/metric.py", line 230, in wrapped_func
    return_value = func(*args, **kwargs)
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/samcli/local/lambdafn/runtime.py", line 178, in invoke
    container = self.create(function_config, debug_context, container_host, container_host_interface)
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/samcli/local/lambdafn/runtime.py", line 396, in create
    self._observer.watch(function_config)
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/samcli/lib/utils/file_observer.py", line 203, in watch
    self._observers[function_config.packagetype].watch(resource)
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/samcli/lib/utils/file_observer.py", line 351, in watch
    self._single_file_observer.watch(resource, self._group)
  File "/usr/local/Cellar/aws-sam-cli/1.37.0/libexec/lib/python3.8/site-packages/samcli/lib/utils/file_observer.py", line 475, in watch
    path_obj = Path(resource)
  File "/usr/local/Cellar/python@3.8/3.8.12_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/pathlib.py", line 1042, in __new__
    self = cls._from_parts(args, init=False)
  File "/usr/local/Cellar/python@3.8/3.8.12_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/pathlib.py", line 683, in _from_parts
    drv, root, parts = self._parse_args(args)
  File "/usr/local/Cellar/python@3.8/3.8.12_1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/pathlib.py", line 667, in _parse_args
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

### Expected result:

SAM initializes Lambda container without error. 

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: macOS Monterey 12.2.1
2. `sam --version`: SAM CLI, version 1.37.0
3. AWS region: us-west-2

**Comment by mildaniel** [comment]
Thanks for submitting this issue. I've confirmed that this is a bug and have submitted a PR to fix it.

**Comment by matteocontrini** [comment]
@mildaniel any insight on when this fix will be released? It makes it quite hard to test locally in an efficient way

**Comment by hooverdc** [comment]
@mildaniel could I ask why the relevant PR for this issue was reverted?

**Comment by mildaniel** [comment]
This fix caused some integration test failures in our canaries. We're looking into resolving it and getting this fix released.

**Comment by qingchm** [comment]
The fix has been released, please update to SAMCLI v1.44.0 for this fix. Thanks for your support and I'm closing this ticket now, please feel free to let us know if you have any further questions!
